### PR TITLE
Basic implementation

### PR DESCRIPTION
**Pull request by regexident**
Hey @rainliu,

this PR adds a basic implementation for WebRTC's data channels, based on Pion's implementation.

Disclaimers:

- Since this crate depends on [sctp](https://github.com/webrtc-rs/sctp), which at this point in time is not yet feature-complete I went ahead and define a few API stubs for the types and functions needed.

- For now I omitted DataChannel's tests, since I'm still rather unfamiliar with Go and Pion's [data-channel tests](https://github.com/pion/datachannel/blob/master/datachannel_test.go) make use of `github.com/pion/transport/test`'s [Bridge](https://github.com/pion/transport/blob/master/test/bridge.go), which makes quite heavy use of Go channels.

I figured however that even at an unfinished state it might be beneficial to open a PR for the stuff that's done.

### CODE REVIEW

**Comment by rainliu** [comment]
Great! Thank you for contributions.

**Comment by rainliu** [comment]
Could you do “cargo fmt” and cargo clippy, test, etc?

**Comment by regexident** [comment]
Done!

**Comment by regexident** [comment]
Interesting. `cargo fmt --all -- --check` succeeds on my checkout. 🤔 

I'm running **rustfmt 1.4.36-stable (7de6968e 2021-02-07)**.

**Comment by rainliu** [comment]
try to update rust to 1.52.1?

**Comment by regexident** [comment]
Weirdly enough I already am on the latest Rust version:

```
$ rustc --version
rustc 1.52.1 (9bc8c42bb 2021-05-09)
```

File: src/channel_type.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+const CHANNEL_TYPE_RELIABLE_UNORDERED: u8 = 0x80;
+const CHANNEL_TYPE_PARTIAL_RELIABLE_REXMIT: u8 = 0x01;
+const CHANNEL_TYPE_PARTIAL_RELIABLE_REXMIT_UNORDERED: u8 = 0x81;
+const CHANNEL_TYPE_PARTIAL_RELIABLE_TIMED: u8 = 0x02;
+const CHANNEL_TYPE_PARTIAL_RELIABLE_TIMED_UNORDERED: u8 = 0x82;
+
+const CHANNEL_TYPE_LEN: usize = 1;
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum Error {
```
Comment (rainliu) [inline src/channel_type.rs:15]: please use thiserror crate for simplifying error handling.

and move all error enums to a dedicated lib/error.rs file.

File: src/data_channel.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+use crate::{
+    exact_size_buf::ExactSizeBuf,
+    marshal::{Marshal, Unmarshal},
+    message::{DataChannelOpen, Error as MessageError, Message, MessageType},
+    sctp::{self, Association, AssociationError, PayloadType, StreamError},
+    ChannelType,
+};
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum Error {
```
Comment (rainliu) [inline src/data_channel.rs:19]: same here fore Error enum.

File: src/sctp.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            todo!()
+        }
+    }
+}
+
+pub use association::{Association, Error as AssociationError};
+
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub enum PayloadType {
+    WebRTCBinary,
```
Comment (rainliu) [inline src/sctp.rs:29]: cargo clippy
warning: name `WebRTCBinary` contains a capitalized acronym
  --> src/sctp.rs:29:5
   |
29 |     WebRTCBinary,
   |     ^^^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `WebRtcBinary`
   |
   = note: `#[warn(clippy::upper_case_acronyms)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

warning: name `WebRTCBinaryEmpty` contains a capitalized acronym
  --> src/sctp.rs:30:5
   |
30 |     WebRTCBinaryEmpty,
   |     ^^^^^^^^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `WebRtcBinaryEmpty`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

warning: name `WebRTCDCEP` contains a capitalized acronym
  --> src/sctp.rs:31:5
   |
31 |     WebRTCDCEP,
   |     ^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `WebRtcdcep`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

warning: name `WebRTCString` contains a capitalized acronym
  --> src/sctp.rs:32:5
   |
32 |     WebRTCString,
   |     ^^^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `WebRtcString`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

warning: name `WebRTCStringEmpty` contains a capitalized acronym
  --> src/sctp.rs:33:5
   |
33 |     WebRTCStringEmpty,
   |     ^^^^^^^^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `WebRtcStringEmpty`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

warning: 5 warnings emitted

**Comment by rainliu** [comment]
it turned out Github Action's toolchain is too old, just upgrade it to 1.52.1

File: src/channel_type.rs (hunk 1 above)
Comment (regexident) [reply src/channel_type.rs]: Done.

File: src/sctp.rs (hunk 3 above)
Comment (regexident) [reply src/sctp.rs]: Fixed.

**Pull request merged**